File: Labs/Lab 7/lab7.java
import java.util.Arrays;

public class lab7 {
    public int findTheCity(int n, int[][] edges, int distanceThreshold) {
        //matrix records shortest distance btwn two cities
        int[][] dis = new int[n][n];
        
        //city w/ the least # of reachable cities
        int res = 0;
        
        //smallest # of reachable cities found (so far)
        int smallest = n;
        
        for (int[] row : dis) //no direct path, distance matrix filled with a large number
            Arrays.fill(row, 10001);
        
        for (int[] e : edges)
            dis[e[0]][e[1]] = dis[e[1]][e[0]] = e[2]; //set the distance for direct edges
        
        
        for (int i = 0; i < n; ++i)
            dis[i][i] = 0; //city and itself distance is 0 
        
        for (int k = 0; k < n; ++k)
            for (int i = 0; i < n; ++i)
                for (int j = 0; j < n; ++j)
                    dis[i][j] = Math.min(dis[i][j], dis[i][k] + dis[k][j]); //Floyd-Warshall alg, compute the shortest paths
        
        for (int i = 0; i < n; i++) {
            int count = 0;
            
            //number of cities within the distance threshold for city 'i'
            for (int j = 0; j < n; ++j)
                if (dis[i][j] <= distanceThreshold)
                    ++count;
            
            //if count<= to the smallest found so far, update the result and smallest count
            if (count <= smallest) {
                res = i;
                smallest = count;
            }
        }
        
        //return city w/ the least # of reachable cities within distance threshold
        return res;
    }
}

File: Labs/Lab 7/lab7.py
from ast import List
import math


class lab7:
    def findTheCity(self, n: int, edges: List[List[int]], distanceThreshold: int) -> int:
    #Your code here
        #matrix w/ distance infinity, set the distance to self as 0 (no self loop for purposes of this class) 
        dist = [[math.inf] * n for _ in range(n)]
        for i in range(n):
            dist[i][i] = 0

        for u, v, w in edges:
            dist[u][v] = dist[v][u] = w #distances filled based on essay 

        #Floyd-Warshall algorithm:distance updated to shortest path so far
        for k in range(n):
            for i in range(n):
                for j in range(n):
                    if dist[i][j] > dist[i][k] + dist[k][j]:
                        dist[i][j] = dist[i][k] + dist[k][j]

        cnt = [0] * n #count cities if within threshold
        for i in range(n):
            for j in range(n):
                if i == j: continue 
                if dist[i][j] <= distanceThreshold:
                    cnt[j] += 1

        #find city with minumum amt of reachable cities
        ans = 0
        for i in range(n):
            if cnt[i] <= cnt[ans]:
                ans = i
        return ans